system/lib/pthread/library_pthread_stub.c

```diff
@@ -389,13 +390,22 @@ void __acquire_ptc() {}
 
 void __release_ptc() {}
 
-// When pthreads is not enabled, we can't use the Atomics futex api to do
-// proper sleeps, so simulate a busy spin wait loop instead.
-void emscripten_thread_sleep(double msecs) {
+// busy spin wait sleep
+static void busy_sleep(double msecs) {
   double start = emscripten_get_now();
   double now = start;
   do {
     _emscripten_yield(now);
     now = emscripten_get_now();
   } while (now - start < msecs);
 }
+
+void __emscripten_atomics_sleep(double);
+
+void emscripten_thread_sleep(double msecs) {
+  if (_emscripten_thread_supports_atomics_wait()) {
+    __emscripten_atomics_sleep(msecs);
+  } else {
+    busy_sleep(msecs);
+  }
```

Comment:
This code calls out to JS from Wasm twice: first to ask if atomics_wait is supported, and then a second time to do the atomic sleep if supported.

I think it would be simpler to move all of this logic into JS side, and do the if()ing there, along with the busy-sleep alternative if atomics sleep is not supported?

Reply:
Or maybe implement `_emscripten_thread_supports_atomics_wait` natively like we do for the threaded case, so it can be very fast: https://github.com/emscripten-core/emscripten/blob/19c52e0f787a6ca2bcd1273faa24b28b74a5eb10/system/lib/pthread/emscripten_thread_state.S#L52-L57

I guess for the single threaded case we would need asm and we could just do:

```
int _emscripten_thread_supports_atomics_wait() {
  static int _supports_atomics_wait = -1;
  if (_supports_atomics_wait == -1) {
     _supports_atomics_wait = _emscripten_thread_supports_atomics_wait_js(); // called just once
  }
  return _supports_atomics_wait;
}
```